### PR TITLE
feat(query): optimize StringSearchLike vector_const

### LINKED ISSUE #8118: Feature: optimize `StringSearchLike`
**Summary**

For tpch q13
```sql

select
    c_count,
    count(*) as custdist
from
    (
        select
            c_custkey,
            count(o_orderkey) as c_count
        from
            customer
                left outer join
            orders
            on c_custkey = o_custkey
                and o_comment not like '%pending%deposits%'
        group by
            c_custkey
    )
        c_orders
group by
    c_count
order by
    custdist desc,
    c_count desc;
```

image

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

split rhs pattern to rhs_vec: vec<string>

judge the lhs string contains the rhs_vec element.

if not contain, return false.

Fixes #8118

### CODE REVIEW

**Comment by TCeason** [comment]
use databend tpch 13 test on my local:

not optimize, first execute:
37 rows in set (0.73 sec)
Read 165000 rows, 10.47 MiB in 0.710 sec., 232.53 thousand rows/sec., 14.76 MiB/sec.


After optimize, first execute:
37 rows in set (0.67 sec)
Read 165000 rows, 10.47 MiB in 0.644 sec., 256.36 thousand rows/sec., 16.27 MiB/sec.


```sql
not optimize, first execute:

mysql> select     c_count,     count(*) as custdist from     (         select             c_custkey,             count(o_orderkey) as c_count         from             customer                 left outer join             orders             on c_custkey = o_custkey
       and o_comment not like '%pending%deposits%'         group by             c_custkey     )         c_orders group by     c_count order by     custdist desc,     c_count desc;
+---------+----------+
| c_count | custdist |
+---------+----------+
|       0 |     5000 |
|      10 |      676 |
|       9 |      651 |
|      11 |      618 |
|      12 |      554 |
|       8 |      548 |
|      13 |      514 |
|       7 |      487 |
|      19 |      485 |
|      18 |      461 |
|      14 |      454 |
|      20 |      444 |
|      16 |      442 |
|      17 |      438 |
|      15 |      430 |
|      21 |      396 |
|      22 |      378 |
|       6 |      355 |
|      23 |      322 |
|      24 |      262 |
|      25 |      188 |
|       5 |      184 |
|      26 |      162 |
|      27 |      138 |
|      28 |      103 |
|       4 |       92 |
|      29 |       59 |
|       3 |       49 |
|      30 |       29 |
|      31 |       26 |
|      32 |       19 |
|       2 |       12 |
|      33 |        8 |
|      34 |        7 |
|      35 |        5 |
|       1 |        3 |
|      36 |        1 |
+---------+----------+
37 rows in set (0.73 sec)
Read 165000 rows, 10.47 MiB in 0.710 sec., 232.53 thousand rows/sec., 14.76 MiB/sec.

After optimize, first execute:

mysql> select     c_count,     count(*) as custdist from     (         select             c_custkey,             count(o_orderkey) as c_count         from             customer                 left outer join             orders             on c_custkey = o_custkey
       and o_comment not like '%pending%deposits%'         group by             c_custkey     )         c_orders group by     c_count order by     custdist desc,     c_count desc;
+---------+----------+
| c_count | custdist |
+---------+----------+
|       0 |     5000 |
|      10 |      676 |
|       9 |      651 |
|      11 |      618 |
|      12 |      554 |
|       8 |      548 |
|      13 |      514 |
|       7 |      487 |
|      19 |      485 |
|      18 |      461 |
|      14 |      454 |
|      20 |      444 |
|      16 |      442 |
|      17 |      438 |
|      15 |      430 |
|      21 |      396 |
|      22 |      378 |
|       6 |      355 |
|      23 |      322 |
|      24 |      262 |
|      25 |      188 |
|       5 |      184 |
|      26 |      162 |
|      27 |      138 |
|      28 |      103 |
|       4 |       92 |
|      29 |       59 |
|       3 |       49 |
|      30 |       29 |
|      31 |       26 |
|      32 |       19 |
|       2 |       12 |
|      33 |        8 |
|      34 |        7 |
|      35 |        5 |
|       1 |        3 |
|      36 |        1 |
+---------+----------+
37 rows in set (0.67 sec)
Read 165000 rows, 10.47 MiB in 0.644 sec., 256.36 thousand rows/sec., 16.27 MiB/sec.

```

File: src/query/functions/src/scalars/comparisons/comparison_like.rs [hunk 1]
```diff
@@ -76,10 +76,25 @@ impl StringSearchImpl for StringSearchLike {
...
+                let mut sub_strings: Vec<&str> =
+                    pattern.split(|c: char| c.is_ascii_punctuation()).collect();
+                sub_strings.retain(|&substring| !substring.is_empty());
                 let re_pattern = like_pattern_to_regex(pattern);
                 let re = BytesRegex::new(&re_pattern)
                     .expect("Unable to build regex from LIKE pattern: {}");
-                BooleanColumn::from_iterator(lhs.scalar_iter().map(|x| op(re.is_match(x))))
+                BooleanColumn::from_iterator(lhs.scalar_iter().map(|x| {
+                    let lhs_str = simdutf8::basic::from_utf8(x)
+                        .expect("Unable to convert lhs value to string: {}");
```
Comment (TCeason) [inline src/query/functions/src/scalars/comparisons/comparison_like.rs:87]: Add a cast and a for loop to reduce a is_match…

File: src/query/functions/src/scalars/comparisons/comparison_like.rs [hunk 2]
```diff
@@ -76,10 +76,25 @@ impl StringSearchImpl for StringSearchLike {
             PatternType::PatternStr => {
                 let pattern = simdutf8::basic::from_utf8(rhs)
                     .expect("Unable to convert the LIKE pattern to string: {}");
+                let mut sub_strings: Vec<&str> =
```
Comment (sundy-li) [inline src/query/functions/src/scalars/comparisons/comparison_like.rs:79]: Other characters in regexp should be taked care:

```
*,+,{},?
```

**Comment by xudong963** [comment]
@mergify update

File: src/query/functions/src/scalars/comparisons/comparison_like.rs (hunk 2 above)
Comment (TCeason) [reply src/query/functions/src/scalars/comparisons/comparison_like.rs:79]: Yep, that's a bug:

```sql
mysql> select * from t;
+------+
| id   |
+------+
| +    |
+------+
1 row in set (0.04 sec)
Read 1 rows, 17.00 B in 0.014 sec., 70.29 rows/sec., 1.17 KiB/sec.

mysql> select * from t where id like '%+%';
Empty set (0.03 sec)
Read 1 rows, 17.00 B in 0.018 sec., 56.15 rows/sec., 954.56 B/sec.

```

**Comment by xudong963** [comment]
I run the ticket with 10G tpch data on q13, the following is flamegraph

f2

**Comment by TCeason** [comment]
@xudong963  I load 10G data on my local pc:

run query

select * from orders where o_comment not like '%pending%deposits%';

in this pr the cost of stringsearchlike split to is_match_at and contains and from_utf8

image

before this pr:

image

**Comment by TCeason** [comment]
15.73s optimize to 11.05s

before optimize:

Database changed
mysql> select count(*) from orders where o_comment not like '%pending%deposits%';
+----------+
| count(*) |
+----------+
| 14838130 |
+----------+
1 row in set (15.73 sec)
Read 15000000 rows, 920.67 MiB in 15.685 sec., 956.31 thousand rows/sec., 58.70 MiB/sec.


after optimize:

Database changed
mysql> select count(*) from orders where o_comment not like '%pending%deposits%';
+----------+
| count(*) |
+----------+
| 14838130 |
+----------+
1 row in set (11.05 sec)
Read 15000000 rows, 920.67 MiB in 10.986 sec., 1.37 million rows/sec., 83.80 MiB/sec.

**Review by xudong963 — APPROVED** [review]

**Comment by TCeason** [comment]
```sql
'root'@mysqldb 11:22:43 [tpch]> select count(*) from orders where o_comment not like '%pending%deposits%';
+----------+
| count(*) |
+----------+
| 14838130 |
+----------+
1 row in set (6.14 sec)
Read 15000000 rows, 920.67 MiB in 6.091 sec., 2.46 million rows/sec., 151.15 MiB/sec.
```

Now the cost is almost on is_match.
image


I try to use [bstr's find](https://docs.rs/bstr/0.1.0/bstr/struct.BStr.html#method.find) to instead of string::from_utf8 + string::find.

And the performance is not good. It cost about 9s.

So I also convert the x to string and then find the required_string.

**Comment by TCeason** [comment]
> ```sql
> 'root'@mysqldb 11:22:43 [tpch]> select count(*) from orders where o_comment not like '%pending%deposits%';
> +----------+
> | count(*) |
> +----------+
> | 14838130 |
> +----------+
> 1 row in set (6.14 sec)
> Read 15000000 rows, 920.67 MiB in 6.091 sec., 2.46 million rows/sec., 151.15 MiB/sec.
> ```
> 
> Now the cost is almost on is_match. image
> 
> I try to use [bstr's find](https://docs.rs/bstr/0.1.0/bstr/struct.BStr.html#method.find) to instead of string::from_utf8 + string::find.
> 
> And the performance is not good. It cost about 9s.
> 
> So I also convert the x to string and then find the required_string.

If the modify is ok, I will modify the function-v2 like in this pr.

**Review by sundy-li — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Comment by TCeason** [comment]
Need modify function-v2 like

**Pull request merged**